File: python/solution_001.py
from config import config
from games.game.game_state import GameState
from copy import deepcopy
from algorithms.ucb_mcts.algorithm import Algorithm as UCBAlgorithm
from computation_graphs.basic_computation_graph.computation_graph import \
    ComputationGraph
from model_builders.model_builder.model_builder import ModelBuilder

import argparse
import logging
import numpy as np
import sys
import tensorflow as tf
import time

logging.basicConfig(
    stream=sys.stdout, level=logging.DEBUG,
    format='%(asctime)s %(message)s', datefmt='%H:%M:%S')
logger = logging.getLogger()

argument_parser = argparse.ArgumentParser()
argument_parser.add_argument('game', type=str)
argument_parser.add_argument('algorithm', type=str)
argument_parser.add_argument('model_builder', type=str)
argument_parser.add_argument('number_of_iterations', type=int)
argument_parser.add_argument('ucb_worker_count', type=int)
argument_parser.add_argument('ucb_strength', type=int)
argument_parser.add_argument('ugtsa_worker_count', type=int)
argument_parser.add_argument('ugtsa_strength', type=int)
argument_parser.add_argument('--debug', action='store_true')
argument_parser.add_argument('--use_initial_state', action='store_true')
args = argument_parser.parse_args()

game_config = config['games'][args.game]

game_state = game_config['game_state']
algorithm_config = game_config['algorithms'][args.algorithm]
algorithm_class = algorithm_config['class']
model_builder = algorithm_config['model_builders'][args.model_builder]

graph = tf.Graph()
model_builder.set_worker_count(args.ugtsa_worker_count)
with graph.as_default():
    model_builder.build()

config = tf.ConfigProto()
config.log_device_placement = args.debug

with tf.Session(config=config, graph=graph) as session:
    session.run(tf.global_variables_initializer())
    for i in range(args.number_of_iterations):
        logger.info('{}: global_step {}'.format(i, session.run(
            graph.get_tensor_by_name('global_step:0'))))

        if args.use_initial_state:
            gs = GameState.random_game_state(game_state)
            if gs.is_final:
                gs.undo_move()
        else:
            gs = deepcopy(game_state)

        computation_graph = ComputationGraph(True)
        transformations = ModelBuilder.create_transformations(
            computation_graph)

        ucb_begin = time.time()
        ucb_algorithm = UCBAlgorithm(
            game_state=deepcopy(gs),
            worker_count=args.ucb_worker_count,
            grow_factor=5,
            exploration_factor=np.sqrt(2),
            removed_root_moves=[])
        for _ in range(args.ucb_strength):
            ucb_algorithm.improve()
        ucb_end = time.time()
        logger.info('{}: UCB took {}'.format(i, ucb_end - ucb_begin))

        ugtsa_begin = time.time()
        ugtsa_algorithm = algorithm_class(
            game_state=deepcopy(gs),
            worker_count=args.ugtsa_worker_count,
            grow_factor=5,
            removed_root_moves=[],
            computation_graph=computation_graph,
            empty_statistic=transformations[0],
            move_rate=transformations[1],
            game_state_as_update=transformations[2],
            updated_statistic=transformations[3],
            updated_update=transformations[4])
        for _ in range(args.ugtsa_strength):
            ugtsa_algorithm.improve()
        ugtsa_end = time.time()
        logger.info('{}: UGTSA took {}'.format(i, ugtsa_end - ugtsa_begin))

        gradients_begin = time.time()
        # zero gradient accumulators
        ModelBuilder.zero_model_gradient_accumulators()
        if args.debug:
            ModelBuilder.model_gradient_accumulators_debug_info()

        # calculate move rate gradients
        move_rates = ugtsa_algorithm.move_rates()

        move_rate_values = []
        ucb_move_rate_values = []
        ugtsa_move_rate_values = []
        for move_rate, ucb_move_rate in zip(
                move_rates, ucb_algorithm.move_rates()):
            move_rate_values += [ugtsa_algorithm.value(move_rate)]
            ucb_move_rate_values += [ucb_algorithm.value(ucb_move_rate)]
            ugtsa_move_rate_values += [
                np.zeros(gs.player_count, dtype=np.float32)]

        loss, move_rates_gradient = ModelBuilder.cost_function(
            move_rate_values, ucb_move_rate_values, ugtsa_move_rate_values)

        logger.info('{}: loss {}'.format(i, loss))
        if args.debug:
            print(move_rates_gradient)

        # accumulate gradients
        computation_graph.model_gradients(
            first_node=0,
            y_grads={
                move_rate: gradient
                for move_rate, gradient in zip(
                    move_rates, move_rates_gradient)})

        # apply gradients
        ModelBuilder.apply_gradients()
        if args.debug:
            ModelBuilder.model_gradient_accumulators_debug_info()
        gradients_end = time.time()
        logger.info('{}: gradients took {}'.format(
            i, gradients_end - gradients_begin))

        # debug
        print([x.number_of_visits for x in ucb_algorithm.tree[:20]])
        print([x.number_of_visits for x in ugtsa_algorithm.tree[:20]])

        if args.debug:
            print(ucb_algorithm.tree[:20])
            print(ugtsa_algorithm.tree[:20])

File: cpp/ugtsa/solution_001.cpp
#include "tensorflow/core/public/session.h"

#include "ugtsa/common.h"
#include "ugtsa/algorithms/ucb_mcts/algorithm.h"
#include "ugtsa/algorithms/computation_graph_mcts/algorithm.h"
#include "ugtsa/computation_graphs/basic_computation_graph/computation_graph.h"
#include "ugtsa/games/omringa/game_state.h"

int main(int argc, char **argv) {
    // omringa__default__big_vertical_lstm_model_builder__10 2 10 300 300 7 7 2 2 30
    // omringa__default__small_vertical_lstm_model_builder__10 2 10 150 150 7 7 2 2 30
    std::string graph_name = argv[1];
    int player_count = std::atoi(argv[2]);
    int worker_count = std::atoi(argv[3]);
    int statistic_size = std::atoi(argv[4]);
    int update_size = std::atoi(argv[5]);
    auto game_state_board_shape = std::vector<int>({std::atoi(argv[6]), std::atoi(argv[7])});
    int game_state_statistic_size = std::atoi(argv[8]);
    int update_statistic_size = std::atoi(argv[9]);
    int seed_size = std::atoi(argv[10]);

    // create session
    tensorflow::Session* session;
    tensorflow::SessionOptions session_options = tensorflow::SessionOptions();
    TF_CHECK_OK(tensorflow::NewSession(session_options, &session));

    // load graph
    common::load_model(session, graph_name);

    for (int i = 0; i < 1; i++) {
        // game state
        auto game_state = games::omringa::GameState();
        // game_state.move_to_random_state();
        // if (game_state.is_final()) {
        //     game_state.undo_move();
        // }

        auto ucb_game_state = std::unique_ptr<games::game::GameState>(game_state.copy());
        auto ugtsa_game_state = std::unique_ptr<games::game::GameState>(game_state.copy());

        // create ucb algorithm
        auto ucb_algorithm = algorithms::ucb_mcts::Algorithm(ucb_game_state.get(), 10, 5, {}, std::sqrt(2.));

        // create ugtsa algorithm
        auto computation_graph = computation_graphs::basic_computation_graph::ComputationGraph(session, "training:0", true);
        auto transformations = common::create_transformations(
            &computation_graph, player_count, worker_count, statistic_size, update_size, game_state_board_shape,
            game_state_statistic_size, update_statistic_size, seed_size);
        auto ugtsa_algorithm = algorithms::computation_graph_mcts::Algorithm(
            ugtsa_game_state.get(), worker_count, 5, {}, &computation_graph, transformations[0], transformations[1],
            transformations[2], transformations[3], transformations[4]);

        for (int i = 0; i < 500; i++) {
            ucb_algorithm.improve();
        }

        for (int i = 0; i < 500; i++) {
            ugtsa_algorithm.improve();
        }

        auto ucb_move_rates = ucb_algorithm.move_rates();
        auto ugtsa_move_rates = ugtsa_algorithm.move_rates();
        std::vector<Eigen::VectorXf, Eigen::aligned_allocator<Eigen::VectorXf>> ucb_move_rate_values;
        std::vector<Eigen::VectorXf, Eigen::aligned_allocator<Eigen::VectorXf>> ugtsa_move_rate_values;
        for (auto &x : ucb_move_rates) ucb_move_rate_values.push_back(ucb_algorithm.value(x));
        for (auto &x : ugtsa_move_rates) ugtsa_move_rate_values.push_back(ugtsa_algorithm.value(x));

        std::cout << "ucb_move_rate_values" << std::endl;
        for (auto &x : ucb_move_rate_values) {
            std::cout << x << std::endl;
        }
        std::cout << "ugtsa_move_rate_values" << std::endl;
        for (auto &x : ugtsa_move_rate_values) {
            std::cout << x << std::endl;
        }

        auto pair = common::cost_function(session, ugtsa_move_rate_values, ucb_move_rate_values, ucb_move_rate_values);
        std::cout << "loss: " << pair.first << std::endl;
        std::cout << "gradients: " << std::endl;
        for (auto & x : pair.second) {
            std::cout << x << std::endl;
        }

        common::zero_model_gradient_accumulators(session);
        computation_graph.accumulate_model_gradients(0, ugtsa_move_rates, pair.second);

        // store model
        common::store_model(session, graph_name + "2");
    }

    // close session
    TF_CHECK_OK(session->Close());

    return 0;
}